### PR TITLE
Minor changes to DirectoryHelper.DeleteDirectory()

### PR DESCRIPTION
**Pull request by nulltoken**
This builds up on #971 and proposes the following modifications

 - Allow derived exceptions from known ones to be whitelisted
 - Trace the number of failed attempts made to delete the directory
 - Use named parameters to make the retry parameters less "magic" to the reader

### CODE REVIEW

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs [hunk 1]
```diff
@@ -85,7 +85,9 @@ private static void TryDeleteDirectory(string directoryPath, int maxAttempts, in
                 }
                 catch (Exception ex)
                 {
-                    if (!whitelist.Contains(ex.GetType()))
+                    var exceptiontype = ex.GetType();
+
+                    if (!whitelist.Select(e => exceptiontype.IsAssignableFrom(e)).Any())
```
Comment (ThomasBarnekow) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Like!

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs [hunk 2]
```diff
@@ -85,7 +85,9 @@ private static void TryDeleteDirectory(string directoryPath, int maxAttempts, in
                 }
                 catch (Exception ex)
                 {
-                    if (!whitelist.Contains(ex.GetType()))
+                    var exceptiontype = ex.GetType();
```
Comment (ThomasBarnekow) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Nitpick: Shouldn't that be ```exceptionType``` ;-)

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 1 above)
Comment (Therzok) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Except it's a verbose way of saying:
`!whitelist.Any(e => exceptiontype.IsAssignableFrom(e))`

Comment (nulltoken) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Fixed

File: LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs (hunk 2 above)
Comment (nulltoken) [inline LibGit2Sharp.Tests/TestHelpers/DirectoryHelper.cs]: Fixed

**Comment by nulltoken** [comment]
@ThomasBarnekow Can you fetch this and make this I didn't break anything, please?

**Comment by ThomasBarnekow** [comment]
@nulltoken Will do.

**Comment by ThomasBarnekow** [comment]
@nulltoken The expression must be turned around like so:
```
if (!whitelist.Any(e => e.IsAssignableFrom(exceptionType)))
```
Otherwise, derived exceptions (e.g., our beloved ```DirectoryNotFoundException```) will be thrown instead of caught.

**Comment by nulltoken** [comment]
>@nulltoken The expression must be turned around like so:

Fixed.

YU SO MEAN IsAssignableFrom()

**Comment by ThomasBarnekow** [comment]
@nulltoken Just ran all the tests. 808 passed. 0 failed. 7 skipped. All good!

I'd have two thoughts which, however, don't stand in the way of merging this fix:
* First, wouldn't it still make sense to catch exceptions in the ```BaseFixture.Dispose()``` method (as ```DeleteDirectory()``` could theoretically throw)? Or should the top-level ```DeleteDirectory()``` method catch unexpected exceptions and write trace messages for these as well? Currently, these unexpected exceptions (if they can ever happen) would lead to failed tests. You might want that but you might as well not.
* Second, would it make sense to rename the top-level ```DeleteDirectory()``` method into ```TryDeleteDirectory()``` as well? My rationale is that it (1) does not ensure directories are indeed deleted and (2) does not provide any direct feedback to the caller (only ```BaseFixture.Dispose()``` at the moment) that it could not delete a directory.

**Comment by nulltoken** [comment]
@ThomasBarnekow Those are good points

>Currently, these unexpected exceptions (if they can ever happen) would lead to failed tests.

I actually think I'd prefer them to fail the test. The noise would raise the chance for a user to open an issue about this and give us an opportunity to analyze/troubleshoot this unexpected situation. 

>Would it make sense to rename the top-level ```DeleteDirectory()``` method into ```TryDeleteDirectory()``` 

Why not? However, by convention, `TryXxxx` methods expose a certain signature (a returned `bool` and an in `out` param). I'd prefer using a name that doesn't conflict with a known pattern.

>Just ran all the tests. 808 passed. 0 failed. 7 skipped. All good!

:cool: Thanks for having checked this :sparkling_heart: I'm going to merge this (after a quick rebase). Let's discuss the `DeleteDirectory` naming proposal in a separate issue.

**Comment by ThomasBarnekow** [comment]
@nulltoken I agree with all of your points.

**Pull request merged**